src/updater/updater.cpp

```diff
@@ -149,446 +95,568 @@ void CheckWeightsConsistency(
+    const auto turn_penalties_index_region =
+        mmapFile(config.turn_penalties_index_path, boost::interprocess::read_only);
 
-                // count statistics for logging
-                counters[fwd_source] += 1;
-                counters[rev_source] += 1;
-            }
-        }); // parallel_for_each
+    // Mapped file pointer for turn indices
+    const extractor::lookup::TurnIndexBlock *turn_index_blocks =
+        reinterpret_cast<const extractor::lookup::TurnIndexBlock *>(
+            turn_penalties_index_region.get_address());
+    BOOST_ASSERT(is_aligned<extractor::lookup::TurnIndexBlock>(turn_index_blocks));
 
-        counters_type merged_counters(num_counters, 0);
-        for (const auto &counters : segment_speeds_counters)
+    std::vector<std::uint64_t> updated_turns;
+    for (std::uint64_t edge_index = 0; edge_index < turn_weight_penalties.size(); ++edge_index)
+    {
+        // Get the turn penalty and update to the new value if required
+        const auto &turn_index = turn_index_blocks[edge_index];
+        auto turn_weight_penalty = turn_weight_penalties[edge_index];
+        auto turn_duration_penalty = turn_duration_penalties[edge_index];
+        if (auto value = turn_penalty_lookup(turn_index))
         {
-            for (std::size_t i = 0; i < counters.size(); i++)
-            {
-                merged_counters[i] += counters[i];
-            }
+            turn_duration_penalty =
+                boost::numeric_cast<TurnPenalty>(std::round(value->duration * 10.));
+            turn_weight_penalty = boost::numeric_cast<TurnPenalty>(std::round(
+                std::isfinite(value->weight) ? value->weight * weight_multiplier
+                                             : turn_duration_penalty * weight_multiplier / 10.));
+
+            turn_duration_penalties[edge_index] = turn_duration_penalty;
+            turn_weight_penalties[edge_index] = turn_weight_penalty;
+            updated_turns.push_back(edge_index);
         }
 
-        for (std::size_t i = 0; i < merged_counters.size(); i++)
+        if (turn_weight_penalty < 0)
         {
-            if (i == LUA_SOURCE)
-            {
-                util::Log() << "Used " << merged_counters[LUA_SOURCE]
-                            << " speeds from LUA profile or input map";
-            }
-            else
-            {
-                // segments_speeds_counters has 0 as LUA, segment_speed_filenames not, thus we need
-                // to susbstract 1 to avoid off-by-one error
-                util::Log() << "Used " << merged_counters[i] << " speeds from "
-                            << config.segment_speed_lookup_paths[i - 1];
-            }
+            util::Log(logWARNING) << "Negative turn penalty at " << turn_index.from_id << ", "
+                                  << turn_index.via_id << ", " << turn_index.to_id
+                                  << ": turn penalty " << turn_weight_penalty;
```

Comment:
So the gist is that we have allowed negative turn penalties for several reason (mostly around the fact that it's easier to define them w.r.t. to the avg speed of the segment), but they are not a good idea and can cause bugs. This warning makes sure we have a paper trail in the logs if someone feeds them in regardless.